// filestack-js/build/main/lib/request/helpers/data.spec.browser.js
"use strict";
/*
 * Copyright (c) 2018 by Filestack
 * Some rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var data_1 = require("./data");
describe('Request/Helpers/Data', function () {
    describe('prepare data', function () {
        it('should return empty data', function () {
            var data = { url: 'https://filestack.com', data: {} };
            expect(data_1.prepareData(data)).toEqual(data);
        });
        it('should return ArrayBuffer', function () {
            var data = { url: 'https://filestack.com', data: new ArrayBuffer(10) };
            expect(data_1.prepareData(data)).toEqual(data);
        });
        it('should return url params', function () {
            var data = { url: 'https://filestack.com', data: new URLSearchParams('q=search&id=1') };
            expect(data_1.prepareData(data)).toEqual(data);
        });
    });
    describe('parse response', function () {
        it('should return equal response data', function () {
            var response = {
                status: 200,
                statusText: 'ok',
                headers: {},
                data: [],
                config: {
                    url: 'https://filestack.com',
                },
            };
            expect(data_1.parseResponse(response)).toEqual(response);
        });
        it('should return response with application/json and data stringify', function () {
            var response = {
                status: 200,
                statusText: 'ok',
                headers: {
                    'content-type': 'application/json',
                },
                data: JSON.stringify({ a: 1 }),
                config: {
                    url: 'https://filestack.com',
                },
            };
            expect(data_1.parseResponse(response)).toEqual(response);
        });
        it('should return response with application/json and json data ', function () {
            var response = {
                status: 200,
                statusText: 'ok',
                headers: {
                    'content-type': 'application/json',
                },
                data: { a: 1 },
                config: {
                    url: 'https://filestack.com',
                },
            };
            expect(data_1.parseResponse(response)).toEqual(response);
        });
        it('should return text/plain response with ArrayBuffer ', function () {
            var response = {
                status: 200,
                statusText: 'ok',
                headers: {
                    'content-type': 'text/plain',
                },
                data: new ArrayBuffer(10),
                config: {
                    url: 'https://filestack.com',
                },
            };
            expect(data_1.parseResponse(response)).toEqual(response);
        });
    });
});
