// backendless/es/files/persmissions/persmission.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _initializerDefineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/initializerDefineProperty"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _applyDecoratedDescriptor2 = _interopRequireDefault(require("@babel/runtime/helpers/applyDecoratedDescriptor"));

var _initializerWarningHelper2 = _interopRequireDefault(require("@babel/runtime/helpers/initializerWarningHelper"));

var _utils = _interopRequireDefault(require("../../utils"));

var _decorators = require("../../decorators");

var _sendRequest = require("./send-request");

var _dec, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _class, _descriptor, _descriptor2, _descriptor3, _descriptor4, _descriptor5, _descriptor6, _descriptor7, _descriptor8, _descriptor9, _descriptor10, _descriptor11, _descriptor12, _temp;

var PermissionTypes = {
  GRANT: 'GRANT',
  DENY: 'DENY'
};

function grantForUser(userId, url, asyncHandler) {
  return _sendRequest.sendRequest.call(this, PermissionTypes.GRANT, url, {
    userId: userId
  }, asyncHandler);
}

function denyForUser(userId, url, asyncHandler) {
  return _sendRequest.sendRequest.call(this, PermissionTypes.DENY, url, {
    userId: userId
  }, asyncHandler);
}

function grantForRole(roleName, url, asyncHandler) {
  return _sendRequest.sendRequest.call(this, PermissionTypes.GRANT, url, {
    roleName: roleName
  }, asyncHandler);
}

function denyForRole(roleName, url, asyncHandler) {
  return _sendRequest.sendRequest.call(this, PermissionTypes.DENY, url, {
    roleName: roleName
  }, asyncHandler);
}

function grantForAllUsers(url, asyncHandler) {
  return _sendRequest.sendRequest.call(this, PermissionTypes.GRANT, url, {
    userId: '*'
  }, asyncHandler);
}

function denyForAllUsers(url, asyncHandler) {
  return _sendRequest.sendRequest.call(this, PermissionTypes.DENY, url, {
    userId: '*'
  }, asyncHandler);
}

function grantForAllRoles(url, asyncHandler) {
  return _sendRequest.sendRequest.call(this, PermissionTypes.GRANT, url, {
    roleName: '*'
  }, asyncHandler);
}

function denyForAllRoles(url, asyncHandler) {
  return _sendRequest.sendRequest.call(this, PermissionTypes.DENY, url, {
    roleName: '*'
  }, asyncHandler);
} //TODO: will be removed when remove sync methods


var namespaceLabel = 'Backendless.Files.Permissions.{READ|DELETE|WRITE}';
var FilePermission = (_dec = (0, _decorators.deprecated)(namespaceLabel, "".concat(namespaceLabel, ".grantForUser")), _dec2 = (0, _decorators.deprecated)(namespaceLabel, "".concat(namespaceLabel, ".grantForUser")), _dec3 = (0, _decorators.deprecated)(namespaceLabel, "".concat(namespaceLabel, ".denyForUser")), _dec4 = (0, _decorators.deprecated)(namespaceLabel, "".concat(namespaceLabel, ".denyForUser")), _dec5 = (0, _decorators.deprecated)(namespaceLabel, "".concat(namespaceLabel, ".grantForRole")), _dec6 = (0, _decorators.deprecated)(namespaceLabel, "".concat(namespaceLabel, ".grantForRole")), _dec7 = (0, _decorators.deprecated)(namespaceLabel, "".concat(namespaceLabel, ".denyForRole")), _dec8 = (0, _decorators.deprecated)(namespaceLabel, "".concat(namespaceLabel, ".denyForRole")), _dec9 = (0, _decorators.deprecated)(namespaceLabel, "".concat(namespaceLabel, ".grantForAllUsers")), _dec10 = (0, _decorators.deprecated)(namespaceLabel, "".concat(namespaceLabel, ".grantForAllUsers")), _dec11 = (0, _decorators.deprecated)(namespaceLabel, "".concat(namespaceLabel, ".denyForAllUsers")), _dec12 = (0, _decorators.deprecated)(namespaceLabel, "".concat(namespaceLabel, ".denyForAllUsers")), (_class = (_temp = function FilePermission(permission, app) {
  (0, _classCallCheck2["default"])(this, FilePermission);
  (0, _initializerDefineProperty2["default"])(this, "grantUserSync", _descriptor, this);
  (0, _initializerDefineProperty2["default"])(this, "grantUser", _descriptor2, this);
  (0, _initializerDefineProperty2["default"])(this, "denyUserSync", _descriptor3, this);
  (0, _initializerDefineProperty2["default"])(this, "denyUser", _descriptor4, this);
  (0, _initializerDefineProperty2["default"])(this, "grantRoleSync", _descriptor5, this);
  (0, _initializerDefineProperty2["default"])(this, "grantRole", _descriptor6, this);
  (0, _initializerDefineProperty2["default"])(this, "denyRoleSync", _descriptor7, this);
  (0, _initializerDefineProperty2["default"])(this, "denyRole", _descriptor8, this);
  (0, _initializerDefineProperty2["default"])(this, "grantSync", _descriptor9, this);
  (0, _initializerDefineProperty2["default"])(this, "grant", _descriptor10, this);
  (0, _initializerDefineProperty2["default"])(this, "denySync", _descriptor11, this);
  (0, _initializerDefineProperty2["default"])(this, "deny", _descriptor12, this);
  (0, _defineProperty2["default"])(this, "grantForUser", _utils["default"].promisified(grantForUser));
  (0, _defineProperty2["default"])(this, "denyForUser", _utils["default"].promisified(denyForUser));
  (0, _defineProperty2["default"])(this, "grantForRole", _utils["default"].promisified(grantForRole));
  (0, _defineProperty2["default"])(this, "denyForRole", _utils["default"].promisified(denyForRole));
  (0, _defineProperty2["default"])(this, "grantForAllUsers", _utils["default"].promisified(grantForAllUsers));
  (0, _defineProperty2["default"])(this, "denyForAllUsers", _utils["default"].promisified(denyForAllUsers));
  (0, _defineProperty2["default"])(this, "grantForAllRoles", _utils["default"].promisified(grantForAllRoles));
  (0, _defineProperty2["default"])(this, "denyForAllRoles", _utils["default"].promisified(denyForAllRoles));
  this.permission = permission;
  this.app = app;
}, _temp), (_descriptor = (0, _applyDecoratedDescriptor2["default"])(_class.prototype, "grantUserSync", [_dec], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: function initializer() {
    return _utils["default"]["synchronized"](grantForUser);
  }
}), _descriptor2 = (0, _applyDecoratedDescriptor2["default"])(_class.prototype, "grantUser", [_dec2], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: function initializer() {
    return _utils["default"].promisified(grantForUser);
  }
}), _descriptor3 = (0, _applyDecoratedDescriptor2["default"])(_class.prototype, "denyUserSync", [_dec3], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: function initializer() {
    return _utils["default"]["synchronized"](denyForUser);
  }
}), _descriptor4 = (0, _applyDecoratedDescriptor2["default"])(_class.prototype, "denyUser", [_dec4], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: function initializer() {
    return _utils["default"].promisified(denyForUser);
  }
}), _descriptor5 = (0, _applyDecoratedDescriptor2["default"])(_class.prototype, "grantRoleSync", [_dec5], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: function initializer() {
    return _utils["default"]["synchronized"](grantForRole);
  }
}), _descriptor6 = (0, _applyDecoratedDescriptor2["default"])(_class.prototype, "grantRole", [_dec6], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: function initializer() {
    return _utils["default"].promisified(grantForRole);
  }
}), _descriptor7 = (0, _applyDecoratedDescriptor2["default"])(_class.prototype, "denyRoleSync", [_dec7], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: function initializer() {
    return _utils["default"]["synchronized"](denyForRole);
  }
}), _descriptor8 = (0, _applyDecoratedDescriptor2["default"])(_class.prototype, "denyRole", [_dec8], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: function initializer() {
    return _utils["default"].promisified(denyForRole);
  }
}), _descriptor9 = (0, _applyDecoratedDescriptor2["default"])(_class.prototype, "grantSync", [_dec9], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: function initializer() {
    return _utils["default"]["synchronized"](grantForAllUsers);
  }
}), _descriptor10 = (0, _applyDecoratedDescriptor2["default"])(_class.prototype, "grant", [_dec10], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: function initializer() {
    return _utils["default"].promisified(grantForAllUsers);
  }
}), _descriptor11 = (0, _applyDecoratedDescriptor2["default"])(_class.prototype, "denySync", [_dec11], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: function initializer() {
    return _utils["default"]["synchronized"](denyForAllUsers);
  }
}), _descriptor12 = (0, _applyDecoratedDescriptor2["default"])(_class.prototype, "deny", [_dec12], {
  configurable: true,
  enumerable: true,
  writable: true,
  initializer: function initializer() {
    return _utils["default"].promisified(denyForAllUsers);
  }
})), _class));
exports["default"] = FilePermission;
